src/Translator.php

```diff
@@ -0,0 +1,542 @@
+            throw new ProcessFailedException($process);
+        }
+        return trim($process->getOutput());
+    }
+
+    private function log(string $message): void
+    {
+        echo "$message\n";
+    }
+
+    /**
+     * Gets the module lang dir
+     */
+    private function getLangDirectory(string $modulePath): string
+    {
+        $sources = $this->getTransifexSources($modulePath);
+        foreach ($sources as $source) {
+            if (preg_match('#^(?<dir>.+)\\/(?<file>[^\\/]+)\\.yml$#', $source, $matches)) {
+                return $modulePath . '/' . $matches['dir'];
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Gets the directories of the JS lang folder.
+     */
+    private function getJSLangDirectories(string $modulePath): array
+    {
+        $sources = $this->getTransifexSources($modulePath);
+        $dirs = [];
+        foreach ($sources as $source) {
+            // Strip out /src/ dir and trailing file.js
+            if (preg_match('#^(?<dir>.+)\\/src\\/(?<file>[^\\/]+)\\.js(on)?$#', $source, $matches)) {
+                $dirs[] = $modulePath . '/' . $matches['dir'];
+            }
+        }
+        return $dirs;
+    }
+
+    /**
+     * Get list of transifex source files. E.g. lang/en.yml
+     */
+    private function getTransifexSources(string $modulePath): array
+    {
+        $path = "$modulePath/.tx/config";
+        $content = file_get_contents($path);
+        $sources = [];
+        foreach (preg_split('~\R~u', $content) as $line) {
+            if (preg_match('#source_file\s=\s(?<path>\S+)#', $line, $matches)) {
+                $sources[] = $matches['path'];
+            }
+        }
+        return $sources;
+    }
+
+    /**
+     * Generate javascript for all modules
+     */
+    private function generateJavascript(): void
+    {
+        $this->log("Generating javascript locale files");
+        // Check which paths in each module require processing
+        $count = 0;
+        foreach ($this->modulePaths as $modulePath) {
+            $jsPaths = $this->getJSLangDirectories($modulePath);
+            foreach ((array)$jsPaths as $jsPath) {
+                $count += $this->generateJavascriptInDirectory($modulePath, $jsPath);
+            }
+        }
+        $this->log("Finished generating {$count} files");
+    }
+
+    /**
+     * Process all javascript in a given path
+     */
+    private function generateJavascriptInDirectory(string $modulePath, string $jsPath): int
+    {
+        $count = 0;
+        foreach (glob("{$jsPath}/src/*.js*") as $sourceFile) {
+            // re-encode contents to ensure they're consistently formatted
+            $sourceContents = $this->jsonEncode($this->jsonDecode(file_get_contents($sourceFile)));
+            $locale = preg_replace('/\.js.*$/', '', basename($sourceFile));
+            $targetFile = dirname(dirname($sourceFile)) . '/' . $locale . '.js';
+            $this->log("Generating file {$targetFile}");
```

Comment:
I don't think we need verbose mode given this is an internal tool. Just output everything.

Reply:
Whether this is an internal tool or not is irrelevant. We should have a verbose mode to avoid unnecessary output that will get in the way of actually seeing what's happening. If I have a terminal full of output for _every single file_ I am liable to simply ignore it, as it will be overwhelming.

Reply:
Updated